src/libtorch.cc

```diff
@@ -0,0 +1,1122 @@
+  // 'instance_group_device_id' to initialize the model when
+  // loading. Return in 'model_path' the full path to the
+  // TorchScript file, return in 'torch_model' the Torch Module
+  // representing the model.
+  TRITONSERVER_Error* LoadModel(
+      const std::string& artifact_name, const torch::Device device,
+      std::string* model_path,
+      std::shared_ptr<torch::jit::script::Module> torch_model);
+
+ private:
+  ModelState(TRITONBACKEND_Model* triton_model);
+  TRITONSERVER_Error* AutoCompleteConfig();
+};
+
+
+TRITONSERVER_Error*
+ModelState::Create(TRITONBACKEND_Model* triton_model, ModelState** state)
+{
+  try {
+    *state = new ModelState(triton_model);
+  }
+  catch (const BackendModelException& ex) {
+    RETURN_ERROR_IF_TRUE(
+        ex.err_ == nullptr, TRITONSERVER_ERROR_INTERNAL,
+        std::string("unexpected nullptr in BackendModelException"));
+    RETURN_IF_ERROR(ex.err_);
+  }
+
+  // Auto-complete the configuration if requested...
+  bool auto_complete_config = false;
+  RETURN_IF_ERROR(TRITONBACKEND_ModelAutoCompleteConfig(
+      triton_model, &auto_complete_config));
+  if (auto_complete_config) {
+    RETURN_IF_ERROR((*state)->AutoCompleteConfig());
+
+    triton::common::TritonJson::WriteBuffer json_buffer;
+    (*state)->ModelConfig().Write(&json_buffer);
+
+    TRITONSERVER_Message* message;
+    RETURN_IF_ERROR(TRITONSERVER_MessageNewFromSerializedJson(
+        &message, json_buffer.Base(), json_buffer.Size()));
+    RETURN_IF_ERROR(TRITONBACKEND_ModelSetConfig(
+        triton_model, 1 /* config_version */, message));
+  }
+
+  return nullptr;  // success
+}
+
+ModelState::ModelState(TRITONBACKEND_Model* triton_model)
+    : BackendModel(triton_model)
+{
+  // TODO Obtain backend configuration and validate
+}
+
+TRITONSERVER_Error*
+ModelState::LoadModel(
+    const std::string& artifact_name, const torch::Device device,
+    std::string* model_path,
+    std::shared_ptr<torch::jit::script::Module> torch_model)
+{
+  // Find the TorchScript file that describes the model. If the model
+  // configuration doesn't have an explicit model file specified then
+  // use the default name ("model.pt").
+  std::string cc_model_filename = artifact_name;
+  if (cc_model_filename.empty()) {
+    cc_model_filename = "model.pt";
+  }
+
+  *model_path = JoinPath(
+      {RepositoryPath(), std::to_string(Version()), cc_model_filename});
+
+  {
+    bool exists;
+    RETURN_IF_ERROR(FileExists(*model_path, &exists));
+    RETURN_ERROR_IF_FALSE(
+        exists, TRITONSERVER_ERROR_UNAVAILABLE,
+        std::string("unable to find '") + *model_path +
+            "' for model instance '" + Name() + "'");
+  }
+
+  // Serialize the torch model to string
+  std::string model_data_str;
+  RETURN_IF_ERROR(ReadTextFile(*model_path, &model_data_str));
+
+  try {
+    std::istringstream model_stream(model_data_str);
+    torch_model = std::make_shared<torch::jit::script::Module>(
```

Comment:
Don't create a new shared (unique) ptr, you juse want to torch_model->reset(...)